sdk/python/lib/pulumi/resource.py

```diff
@@ -49,6 +53,160 @@ def __init__(self,
+    """
+
+#   If the child name has the parent name as a prefix, then we make the assumption that it was
+#   constructed from the convention of using `{name}-details` as the name of the child resource.  To
+#   ensure this is aliased correctly, we must then also replace the parent aliases name in the prefix of
+#   the child resource name.
+#
+#   For example:
+#   * name: "newapp-function"
+#   * opts.parent.__name: "newapp"
+#   * parentAlias: "urn:pulumi:stackname::projectname::awsx:ec2:Vpc::app"
+#   * parentAliasName: "app"
+#   * aliasName: "app-function"
+#   * childAlias: "urn:pulumi:stackname::projectname::aws:s3/bucket:Bucket::app-function"
+    from .output import Output as Op
+    alias_name = Op.from_input(child_name)
+    if child_name.startswith(parent_name):
+        alias_name = Op.from_input(parent_alias).apply(
+            lambda u: u[u.rfind("::") + 2:] + child_name[len(parent_name):])
+
+    return create_urn(alias_name, child_type, parent_alias)
+
+ABSENT_VALUE = object()
+"""
+ABSENT_VALUE is how we mark if a value provided to an alias was actually given by the user as 'None',
+versus the use not providing any value at all.
+"""
+
+ROOT_STACK_RESOURCE = None
+"""
+Constant to represent the 'root stack' resource for a Pulumi application.  The purpose of this is
+solely to make it easy to write an [Alias] like so:
+
+`aliases=[Alias(parent=pulumi.ROOT_STACK_RESOURCE)]`.
+
+This indicates that the prior name for a resource was created based on it being parented directly by
+the stack itself and no other resources.  Note: this is equivalent to:
+
+`aliases=[Alias(parent=None)]`
+
+However, the former form is preferable as it is more self-descriptive, while the latter may look a
+bit confusing and may incorrectly look like something that could be removed without changing
+semantics.
+"""
+
+class Alias:
+    """
+    Alias is a partial description of prior named used for a resource. It can be processed in the
+    context of a resource creation to determine what the full aliased URN would be.
+
+    Note there is a semantic difference between attributes being given the `None` value and
+    attributes not being given at all. Specifically, there is a difference between:
+
+    ```ts
+    Alias(name="foo", parent=None) # and
+    Alias(name="foo")
+    ```
+
+    So the first alias means "the original urn had no parent" while the second alias means "use the
+    current parent".
+
+    Note: to indicate that a resource was previously parented by the root stack, it is recommended
+    that you use:
+
+    `aliases=[Alias(parent=pulumi.ROOT_STACK_RESOURCE)]`
+
+    This form is self-descriptive and makes the intent clearer than using:
+
+    `aliases=[Alias(parent=None)]`
+    """
+
+    name: Optional[str]
+    """
+    The previous name of the resource.  If not provided, the current name of the resource is used.
+    """
+
+    typ: Optional[str]
```

Comment:
Elsewhere we’ve used postfix `_` when there are conflicts with legal names - which i believe is the standard python naming guidance. We should probably be consistent  here.

Reply:
sure.